### PR TITLE
tests: share the cache in ros tests

### PR DESCRIPTION
**Pull request by elopio**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

### CODE REVIEW

**Comment by kyrofa** [comment]
What about the tests that aren't in the snapd suite? They don't all strictly need to run (although it wouldn't hurt, it would slow things down further).

**Comment by elopio** [comment]
@kyrofa we can turn this into a ros shared cache fixture. It could be a base class, but then we end up with multiple inheritance which makes me nervous.

**Comment by sergiusens** [comment]
Yes, I would say no on multiple inheritance as well

File: snapcraft/tests/fixture_setup.py [hunk 1]
```diff
@@ -1013,3 +1013,23 @@ def _stop_fake_server(self, thread):
...
+class SharedCache(fixtures.Fixture):
+
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def setUp(self):
+        super().setUp()
+        shared_cache_dir = os.path.join(
+            tempfile.gettempdir(), 'snapcraft_test_cache_{}'.format(self.name))
```
Comment (sergiusens) [inline snapcraft/tests/fixture_setup.py]: This makes it so that there is no share cache

Comment (sergiusens) [reply snapcraft/tests/fixture_setup.py]: Oh, my mistake, just saw what `tempfile.gettempdir` is doing. Why not set it the the actual `xdg_cache_home`?

Comment (elopio) [reply snapcraft/tests/fixture_setup.py]: The xdg_cache_home is patched in all the tests. I started complicating things trying to make an exception to that rule only for ros on the other PR. So, I jumped back a little and thought that we don't really want to pollute home in any test. This sounded like a clean and better approach to me.

**Comment by sergiusens** [comment]
On vie, nov 10, 2017 at 12:58 PM, Leo Arias <notifications@github.com> 
wrote:
> @elopio commented on this pull request.
> 
> In snapcraft/tests/fixture_setup.py:
> 
> > @@ -1013,3 +1013,23 @@ def _stop_fake_server(self, thread):
>          self.server.shutdown()
>          self.server.socket.close()
>          thread.join()
> +
> +
> +class SharedCache(fixtures.Fixture):
> +
> +    def __init__(self, name):
> +        super().__init__()
> +        self.name = name
> +
> +    def setUp(self):
> +        super().setUp()
> +        shared_cache_dir = os.path.join(
> +            tempfile.gettempdir(), 
> 'snapcraft_test_cache_{}'.format(self.name))
> The xdg_cache_home is patched in all the tests. I started 
> complicating things trying to make an exception to that rule only for 
> ros on the other PR. So, I jumped back a little and thought that we 
> don't really want to pollute home in any test. This sounded like a 
> clean and better approach to me.
> 

But you are using tmp! Maybe better use /var/tmp
>

**Comment by elopio** [comment]
@sergiusens I'm using whatever tempdir returns. Instead of hardcoding /var/tmp, I prefer to pass it in an env var:
```
>>> import os
>>> os.environ['TMPDIR'] = '/var/tmp'
>>> import tempfile
>>> tempfile.gettempdir()
'/var/tmp'
```

That way the caller of the test can control the tmp to use.

**Comment by kyrofa** [comment]
@elopio do you plan on adding that variable definition into the autopkgtest, then?

**Comment by elopio** [comment]
@kyrofa it's already in the integration tests: https://github.com/snapcore/snapcraft/blob/master/debian/tests/integrationtests#L19
I'm not sure if we should add it to the snaps tests too, because we want to remove the ros tests from there soon.

**Comment by kyrofa** [comment]
Nah, works for me.

**Pull request merged**